qt, refactor: Make BitcoinUnits::Unit a scoped enum

hebasto opened the pull request:
Since Qt 5.5 there are [means](https://doc.qt.io/qt-5/qobject.html#Q_ENUM) to register an enum type with the meta-object system (such enum still lacks an ability to interact with [`QSettings::setValue()`](https://doc.qt.io/qt-5/qsettings.html#setValue) and [`QSettings::value()`](https://doc.qt.io/qt-5/qsettings.html#value) without defined stream operators).

In order to reduce global namespace polluting and to force strong type checking, this PR:
- makes `BitcoinUnits::SeparatorStyle` a scoped enum (done in https://github.com/bitcoin-core/gui/pull/3)
- makes `BitcoinUnits::Unit` a scoped enum (`typedef BitcoinUnits::Unit BitcoinUnit;`)
- cleans `BitcoinUnits` class and its functions
- does not change behavior

promag commented on src/qt/bitcoinunits.cpp:
```diff
@@ -6,100 +6,83 @@
+    case Unit::uBTC: return QString::fromUtf8("µBTC (bits)");
+    case Unit::SAT: return QString("Satoshi (sat)");
+    } // no default case, so the compiler can warn about missing cases
+    assert(false);
```
In Qt there's `Q_UNREACHABLE()`.

hebasto replied:
Correct. But I'd prefer to stick to our [Developer Notes](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#c-data-structures) ;)
When C++ will provide the standard way to handle such cases, it will be easier to implement the appropriate change.

Or did I miss something important why `Q_UNREACHABLE()` should be preferable in Qt code?

promag replied:
No preference!

promag reviewed (commented):
Concept ACK, and change LGTM.

practicalswift commented:
Concept ACK: structured is better than unstructured :)

elichai commented:
Concept ACK. will review later this week

jonasschnelli commented:
Concept ACK.
Flew over the code (cf9507442fbec9252f557ed0de8237b631f89d39),... looks good. Needs testing.

Empact commented:
Concept ACK

hebasto commented:
Rebased after #17453 has been merged.

hebasto commented:
Rebased 22be3d60cab774ffd2f8a287f522176b5a0abdfe -> 4ead3f47c35b40382007fd0eee2b01c29581d1e0 due to the conflict with #18278.

jb55 commented on src/qt/transactiontablemodel.h:
```diff
@@ -101,7 +101,7 @@ class TransactionTableModel : public QAbstractTableModel
     QString formatTxType(const TransactionRecord *wtx) const;
     QString formatTxToAddress(const TransactionRecord *wtx, bool tooltip) const;
-    QString formatTxAmount(const TransactionRecord *wtx, bool showUnconfirmed=true, BitcoinUnits::SeparatorStyle separators=BitcoinUnits::separatorStandard) const;
+    QString formatTxAmount(const TransactionRecord* wtx, bool showUnconfirmed = true, BitcoinUnits::SeparatorStyle separators = BitcoinUnits::SeparatorStyle::STANDARD) const;
```
was `const TransactionRecord *wtx` -> `const TransactionRecord* wtx` intentional? looks inconsistent wrt surrounding code

hebasto replied:
[`clang-format-diff.py`](https://github.com/bitcoin/bitcoin/tree/master/contrib/devtools#clang-format-diffpy) was applied.

hebasto commented:
Rebased 4ead3f47c35b40382007fd0eee2b01c29581d1e0 -> cca2ebdaf7f9af26a3381273f79efd9fb95ed1a9  ([pr17877.03](https://github.com/hebasto/bitcoin/commits/pr17877.03) -> [pr17877.04](https://github.com/hebasto/bitcoin/commits/pr17877.04)) due to the conflict with #17905.

hebasto commented:
Rebased cca2ebdaf7f9af26a3381273f79efd9fb95ed1a9 -> a2edcd60a614053ec642f21cedf4858b650c726b ([pr17877.04](https://github.com/hebasto/bitcoin/commits/pr17877.04) -> [pr17877.05](https://github.com/hebasto/bitcoin/commits/pr17877.05)) due to the conflict with #16528.

hebasto commented:
Rebased a2edcd60a614053ec642f21cedf4858b650c726b -> 61c84eab0c5451b2a19b51047c493b45d4562eff ([pr17877.05](https://github.com/hebasto/bitcoin/commits/pr17877.05) -> [pr17877.06](https://github.com/hebasto/bitcoin/commits/pr17877.06)) due to the conflict with #16432.

MarcoFalke commented:
Can't this be done with a scripted diff like in fa5997bd6f?

hebasto commented:
Updated 61c84eab0c5451b2a19b51047c493b45d4562eff -> 53bf585eb918f1674febcb0d2b7ece635d2cc767 ([pr17877.06](https://github.com/hebasto/bitcoin/commits/pr17877.06) -> [pr17877.07](https://github.com/hebasto/bitcoin/commits/pr17877.07), [diff](https://github.com/hebasto/bitcoin/compare/pr17877.06..pr17877.07)):
- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/17877#issuecomment-636351875):

> Can't this be done with a scripted diff like in [fa5997b](https://github.com/bitcoin/bitcoin/commit/fa5997bd6fc82e16b597ea96e3c5c665f1f174ab)?

The `SeparatorStyle` enum has been refactored in a scripted diff. Skipped for the `Unit` enum as that changes touch a way different things.

MarcoFalke commented on src/qt/bitcoinamountfield.cpp:
```diff
@@ -326,14 +326,14 @@ void BitcoinAmountField::unitChanged(int idx)
 
     // Determine new unit ID
-    int newUnit = unit->itemData(idx, BitcoinUnits::UnitRole).toInt();
+    BitcoinUnit newUnit = unit->itemData(idx, BitcoinUnits::UnitRole).value<BitcoinUnit>();
```
Why is this safe to do in qt? Any integer can be converted to `BitcoinUnit`, at least in C++.

hebasto replied:
I believe, no `int` to `BitcoinUnit` conversion here. [`QComboBox::itemData()`](https://doc.qt.io/qt-5/qcombobox.html#itemData) does a kind of mapping from `int` index to `QVariant`.

Anyway, going to make this code more safer.

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/17877#issuecomment-636514466).

MarcoFalke commented on src/qt/optionsmodel.cpp:
```diff
@@ -68,9 +68,10 @@ void OptionsModel::Init(bool resetSettings)
-        settings.setValue("nDisplayUnit", BitcoinUnits::BTC);
-    nDisplayUnit = settings.value("nDisplayUnit").toInt();
+        settings.setValue("nDisplayUnit", static_cast<int>(BitcoinUnit::BTC));
+    nDisplayUnit = static_cast<BitcoinUnit>(settings.value("nDisplayUnit").toInt());
```
Same here

hebasto replied:
See https://github.com/bitcoin/bitcoin/pull/17877#issuecomment-636514466

MarcoFalke commented on src/qt/bitcoinunits.h:
```diff
@@ -59,8 +58,6 @@ class BitcoinUnits: public QAbstractListModel
     //! Get list of units, for drop-down box
     static QList<Unit> availableUnits();
-    //! Is unit ID valid?
-    static bool valid(Unit unit);
```
Same

hebasto replied:
See https://github.com/bitcoin/bitcoin/pull/17877#issuecomment-636514466

MarcoFalke commented on src/qt/bitcoinunits.cpp:
```diff
@@ -26,60 +26,57 @@ QList<BitcoinUnit> BitcoinUnits::availableUnits()
     case Unit::mBTC: return QString("mBTC");
     case Unit::uBTC: return QString::fromUtf8("µBTC (bits)");
     case Unit::SAT: return QString("Satoshi (sat)");
-    default: return QString("???");
```
Same

hebasto replied:
See https://github.com/bitcoin/bitcoin/pull/17877#issuecomment-636514466

MarcoFalke commented:
ACK f8cd12dfd39104d875a16ea0226b4be1450e77f4

hebasto commented:
Updated 53bf585eb918f1674febcb0d2b7ece635d2cc767 -> 25df40fcadaf5b284e94765e25213eb224b33a25 ([pr17877.07](https://github.com/hebasto/bitcoin/commits/pr17877.07) -> [pr17877.08](https://github.com/hebasto/bitcoin/commits/pr17877.08), [diff](https://github.com/hebasto/bitcoin/compare/pr17877.07..pr17877.08)):
- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/17877#discussion_r432940791):

> Why is this safe to do in qt? Any integer can be converted to `BitcoinUnit`, at least in C++.

---

For more type safety I suggest in a follow-up to implement the `operator<<()` and `operator>>()` for `BitcoinUnit` type, and use `QVariant` instances to set and read values from the `QSettings` object directly.

MarcoFalke commented:
Still ACK on the first commit, but I don't feel comfortable about the rest. Previously a corrupt settings file (or a file from a previous version) would recover with a gui glitch. Now it crashes the node.

hebasto committed:
qt: Use QVariant instead of int for BitcoinUnit in QSettings

This change improves type safety.

hebasto commented:
Updated 25df40fcadaf5b284e94765e25213eb224b33a25 -> 949669431de3ecc41d16e58d323dad741588bf79 ([pr17877.08](https://github.com/hebasto/bitcoin/commits/pr17877.08) -> [pr17877.09](https://github.com/hebasto/bitcoin/commits/pr17877.09), [diff](https://github.com/hebasto/bitcoin/compare/pr17877.08..pr17877.09)):

- added _"qt: Use QVariant instead of int for BitcoinUnit in QSettings"_ commit to address @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/17877#issuecomment-636533636):

> Still ACK on the first commit, but I don't feel comfortable about the rest. Previously a corrupt settings file (or a file from a previous version) would recover with a gui glitch.

---

> Now it crashes the node.

Mind providing the `Bitcoin-Qt.conf` file and steps to reproduce a crash?

MarcoFalke commented:
> Mind providing the Bitcoin-Qt.conf file and steps to reproduce a crash?

Sorry I suspected that this crashes the node based on reading the code, I haven't actually tested or verified that.

MarcoFalke commented:
And with the new code, `FromQuint8` seems to deal with the issue.

hebasto commented:
This variant is backward compatible as it uses a new key "display_unit" in `QSettings` file.

hebasto commented:
Rebased 949669431de3ecc41d16e58d323dad741588bf79 -> cbd498e593498d3b28c0c72032f1e62a2f9ef2df ([pr17877.09](https://github.com/hebasto/bitcoin/commits/pr17877.09) -> [pr17877.10](https://github.com/hebasto/bitcoin/commits/pr17877.10)) due to the conflict with #19104.

MarcoFalke commented:
Maybe split out the scripted-diff into a separate pull request: https://github.com/bitcoin-core/gui/pulls

It seems that it can be reviewed by anyone familiar with C++. As opposed to the latter commits, which require specific gui knowledge.

hebasto commented:
@MarcoFalke 
> Maybe split out the scripted-diff into a separate pull request: https://github.com/bitcoin-core/gui/pulls
> 
> It seems that it can be reviewed by anyone familiar with C++.

Done: https://github.com/bitcoin-core/gui/pull/3

> As opposed to the latter commits, which require specific gui knowledge.

Should they be also moved to https://github.com/bitcoin-core/gui ?

hebasto commented:
Rebased cbd498e593498d3b28c0c72032f1e62a2f9ef2df -> 1a348c8ceddcb38ba33caf3ceade419c7590c879 ([pr17877.10](https://github.com/hebasto/bitcoin/commits/pr17877.10) -> [pr17877.11](https://github.com/hebasto/bitcoin/commits/pr17877.11)) due to the merging of https://github.com/bitcoin-core/gui/pull/3.

hebasto commented:
Rebased 1a348c8ceddcb38ba33caf3ceade419c7590c879 -> 72819bd15cb8c77fc6e1162c046060522b3ce6e5 ([pr17877.11](https://github.com/hebasto/bitcoin/commits/pr17877.11) -> [pr17877.12](https://github.com/hebasto/bitcoin/commits/pr17877.12)) due to the conflict with #19314.

hebasto committed:
refactor: Make BitcoinUnits::Unit a scoped enum

hebasto committed:
refactor: Remove BitcoinUnits::valid() function

Since BitcoinUnits::Unit became a scoped enum, BitcoinUnits::valid()
gets needless.

hebasto committed:
refactor: Remove default cases for scoped enum

hebasto commented:
Updated 72819bd15cb8c77fc6e1162c046060522b3ce6e5 -> 1575ffbbe42dd6c78b131dffe8cb00f636141cc1 ([pr17877.12](https://github.com/hebasto/bitcoin/commits/pr17877.12) -> [pr17877.13](https://github.com/hebasto/bitcoin/commits/pr17877.13), [diff](https://github.com/hebasto/bitcoin/compare/pr17877.12..pr17877.13)):

- fixed the conflict with #18027

hebasto commented:
Rebased 1575ffbbe42dd6c78b131dffe8cb00f636141cc1 -> 243205f85b8b10e4f43865557cdd8ea3aa71c380 ([pr17877.13](https://github.com/hebasto/bitcoin/commits/pr17877.13) -> [pr17877.14](https://github.com/hebasto/bitcoin/commits/pr17877.14)) due to the conflict with #19011.

fanquake commented:
Lots of Concept ACKs, but not really any review in 6 months. Lets try and kick-start this over on the [GUI repo](https://github.com/bitcoin-core/gui).

hebasto commented:
Moved to https://github.com/bitcoin-core/gui/pull/60.
